website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,334 @@
+
+    addonWarning: 'The following addons will be effected by this change.',
+};
+
+function getNodesOriginal(nodeTree, nodesOriginal) {
+    /**
+     * take treebeard tree structure of nodes and get a dictionary of parent node and its
+     * children
+     */
+    var i;
+    var nodeId = nodeTree.node.id;
+    nodesOriginal[nodeId] = {
+        public: nodeTree.node.is_public,
+        id: nodeTree.node.id,
+        addons: nodeTree.node.addons,
+        title: nodeTree.node.title,
+        changed: false
+    };
+
+    if (nodeTree.children) {
+        for (i in nodeTree.children) {
+            nodesOriginal = getNodesOriginal(nodeTree.children[i], nodesOriginal);
+        }
+    }
+    return nodesOriginal;
+}
+
+//function patchNodesPrivacy(nodes) {
+//    /**
+//     * patche all the nodes in a changed state.
+//     * we will use API V2 bulk requests when it becomes available
+//     */
+//    var promise;
+//    for (var key in nodes) {
+//        var node = nodes[key];
+//        var url = API_BASE + node.id + '/';
+//        nodesPatch = []
+//        nodesPatch.push({
+//        JSON.stringify(
+//            {
+//                'data': {
+//                    'type': 'nodes',
+//                    'id': node.id,
+//                    'attributes': {
+//                        'public': node.public
+//                    }
+//                }
+//            })}
+//        });
+//    }
+//
+//        promise = $.when(promise,
+//            $.ajax({
+//                url: url,
+//                type: 'PATCH',
+//                dataType: 'json',
+//                contentType: 'application/json',
+//                crossOrigin: true,
+//                xhrFields: {withCredentials: true},
+//                processData: false,
+//                data: JSON.stringify(
+//                    {
+//                        'data': {
+//                            'type': 'nodes',
+//                            'id': node.id,
+//                            'attributes': {
+//                                'public': node.public
+//                            }
+//                        }
+//                    })
+//            }).done(function (response) {
+//            }).fail(function (xhr, status, error) {
+//                $osf.growl('Error', 'Unable to update Projects and/or Components');
+//                Raven.captureMessage('Could not PATCH project settings.', {
+//                    url: url, status: status, error: error
+//                });
+//            }));
+//
+//
+//    return promise;
+//
+//}
+
+function patchNodesPrivacy(nodes) {
+    /**
+     * patche all the nodes in a changed state.
+     * we will use API V2 bulk requests when it becomes available
+     */
+    var promise;
+    for (var key in nodes) {
+        var node = nodes[key];
+        var url = API_BASE + node.id + '/';
+
+        promise = $.when(promise,
```

Comment:
Rather than reassigning `promise` and calling `when` multiple times, it would be call `when` an array of promises (jqXHRs).

```js
var requests = [..build up array of requests];
var promise = $.when(requests);
```

Caveat: I can't tell from the jQuery docs if $.when can actually take an Array. It's worth trying though. If not, look into  `Function.prototype.apply` and see if you can use it with `$.when`.